In [1]:
import pandas as pd

In [2]:
df_events = pd.read_csv("fact_events.csv")
df_events.head()

,event_id,store_id,campaign_id,product_code,base_price(before_promo),quantity_sold(before_promo),promo_type,base_price(after_promo),quantity_sold(after_promo)
0,8481be,STCHE-1,CAMP_DIW_01,P04,290,327.0,25% OFF,217,287
1,20618e,STCHE-3,CAMP_SAN_01,P04,370,379.0,BOGOF,185,1622
2,f30579,STBLR-9,CAMP_DIW_01,P02,860,337.0,33% OFF,576,488
3,4f570c,STBLR-7,CAMP_DIW_01,P05,55,122.0,25% OFF,41,107
4,6d153f,STHYD-5,CAMP_SAN_01,P15,3000,122.0,500 Cashback,2500,272


In [3]:
df_events.shape

(1510, 9)

In [4]:
duplicated_df = df_events[df_events.duplicated()]
duplicated_df

,event_id,store_id,campaign_id,product_code,base_price(before_promo),quantity_sold(before_promo),promo_type,base_price(after_promo),quantity_sold(after_promo)
74,46d57b,STBLR-9,CAMP_DIW_01,P13,350,94.0,BOGOF,175,329
410,f6aa36,STVSK-3,CAMP_DIW_01,P06,415,63.0,25% OFF,311,54
541,6.24E+11,STMLR-0,CAMP_SAN_01,P03,200,206.0,BOGOF,100,541
759,0f8686,STVSK-0,CAMP_SAN_01,P07,300,24.0,BOGOF,150,92
922,491ff2,STVJD-1,CAMP_SAN_01,P12,62,30.0,50% OFF,31,42
1003,4ad12b,STBLR-4,CAMP_DIW_01,P15,3000,407.0,500 Cashback,2500,1245
1155,cd4c5a,STMLR-0,CAMP_SAN_01,P09,90,34.0,25% OFF,67,28
1227,0491f4,STVSK-2,CAMP_DIW_01,P11,190,38.0,50% OFF,95,58
1313,5c7b20,STCBE-4,CAMP_DIW_01,P05,55,101.0,25% OFF,41,89
1401,d622c3,STHYD-0,CAMP_SAN_01,P10,50,25.0,25% OFF,37,20


In [5]:
df_events1 = df_events.drop_duplicates(subset=['store_id', 'campaign_id','product_code'])
df_events1.shape

(1500, 9)

1. The operations team wants to ensure the integrity of the events data by removing duplicates. Check for and remove duplicate rows in the events dataframe based on store_id, campaign_id, and product_code. How many duplicate rows were removed?
- 10 rows were removed

In [6]:
df_stores=pd.read_csv("dim_stores.csv")
df_stores.head()

,store_id,city
0,STTRV-0,Trivandrum
1,STMDU-3,Madurai
2,STHYD-6,Hyderabad
3,STVSK-1,Visakhapatnam
4,STCBE-3,Coimbatore


In [7]:
result = df_stores.groupby('city')['store_id'].count()
result = result[result > 5]

print(result)

city
Bengaluru    10
Chennai       8
Hyderabad     7
Name: store_id, dtype: int64


2. How many cities have more than 5 stores?
- 3 cities have more than 5 stores

In [8]:
df_events1.isnull().sum()

event_id                        0
store_id                        0
campaign_id                     0
product_code                    0
base_price(before_promo)        0
quantity_sold(before_promo)    20
promo_type                      0
base_price(after_promo)         0
quantity_sold(after_promo)      0
dtype: int64

In [9]:
med=df_events1['quantity_sold(before_promo)'].median()
med

np.float64(78.0)

In [10]:
df_events1['quantity_sold(before_promo)']=df_events1['quantity_sold(before_promo)'].fillna(med)

C:\Users\Tanmay\AppData\Local\Temp\ipykernel_24400\1521405841.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_events1['quantity_sold(before_promo)']=df_events1['quantity_sold(before_promo)'].fillna(med)


In [11]:
df_events1.isnull().sum()

event_id                       0
store_id                       0
campaign_id                    0
product_code                   0
base_price(before_promo)       0
quantity_sold(before_promo)    0
promo_type                     0
base_price(after_promo)        0
quantity_sold(after_promo)     0
dtype: int64

3. The sales team has identified missing values in the quantity_sold(before_promo) data. Estimate these values using the median quantity sold before the promotion. How many missing values were filled, and what is the median used for imputation?
- Total 20 missing values were filled. 78 is the median used for imputation

In [12]:
df_products = pd.read_csv('dim_products.csv')
df_products.head()

,product_code,product_name,category
0,P01,Atliq_Masoor_Dal (1KG),Grocery & Staples
1,P02,Atliq_Sonamasuri_Rice (10KG),Grocery & Staples
2,P03,Atliq_Suflower_Oil (1L),Grocery & Staples
3,P04,Atliq_Farm_Chakki_Atta (1KG),Grocery & Staples
4,P05,Atliq_Scrub_Sponge_For_Dishwash,Home Care


In [13]:
df_events1.head()

,event_id,store_id,campaign_id,product_code,base_price(before_promo),quantity_sold(before_promo),promo_type,base_price(after_promo),quantity_sold(after_promo)
0,8481be,STCHE-1,CAMP_DIW_01,P04,290,327.0,25% OFF,217,287
1,20618e,STCHE-3,CAMP_SAN_01,P04,370,379.0,BOGOF,185,1622
2,f30579,STBLR-9,CAMP_DIW_01,P02,860,337.0,33% OFF,576,488
3,4f570c,STBLR-7,CAMP_DIW_01,P05,55,122.0,25% OFF,41,107
4,6d153f,STHYD-5,CAMP_SAN_01,P15,3000,122.0,500 Cashback,2500,272


In [14]:
df_merged=df_products.merge(df_events, on='product_code')
df_merged.head()

,product_code,product_name,category,event_id,store_id,campaign_id,base_price(before_promo),quantity_sold(before_promo),promo_type,base_price(after_promo),quantity_sold(after_promo)
0,P01,Atliq_Masoor_Dal (1KG),Grocery & Staples,31c430,STHYD-1,CAMP_SAN_01,172,312.0,33% OFF,115,393
1,P01,Atliq_Masoor_Dal (1KG),Grocery & Staples,029ff8,STMLR-0,CAMP_SAN_01,172,169.0,33% OFF,115,236
2,P01,Atliq_Masoor_Dal (1KG),Grocery & Staples,fad26a,STCHE-3,CAMP_DIW_01,172,369.0,33% OFF,115,546
3,P01,Atliq_Masoor_Dal (1KG),Grocery & Staples,166989,STCBE-1,CAMP_SAN_01,172,169.0,33% OFF,115,206
4,P01,Atliq_Masoor_Dal (1KG),Grocery & Staples,3cb389,STCHE-6,CAMP_DIW_01,172,309.0,33% OFF,115,460


In [15]:
min_price = df_merged['base_price(before_promo)'].min()

df_merged[df_merged['base_price(before_promo)'] == min_price]

,product_code,product_name,category,event_id,store_id,campaign_id,base_price(before_promo),quantity_sold(before_promo),promo_type,base_price(after_promo),quantity_sold(after_promo)
704,P10,Atliq_Cream_Beauty_Bathing_Soap (125GM),Personal Care,ed7b12,STVSK-0,CAMP_SAN_01,50,16.0,25% OFF,37,13
706,P10,Atliq_Cream_Beauty_Bathing_Soap (125GM),Personal Care,92151c,STMDU-3,CAMP_SAN_01,50,25.0,25% OFF,37,20
708,P10,Atliq_Cream_Beauty_Bathing_Soap (125GM),Personal Care,decd5e,STMDU-2,CAMP_SAN_01,50,21.0,25% OFF,37,17
709,P10,Atliq_Cream_Beauty_Bathing_Soap (125GM),Personal Care,67820e,STBLR-0,CAMP_SAN_01,50,31.0,25% OFF,37,25
715,P10,Atliq_Cream_Beauty_Bathing_Soap (125GM),Personal Care,fedf26,STMLR-2,CAMP_SAN_01,50,19.0,25% OFF,37,15
716,P10,Atliq_Cream_Beauty_Bathing_Soap (125GM),Personal Care,d31787,STMYS-2,CAMP_SAN_01,50,NaN,25% OFF,37,20
719,P10,Atliq_Cream_Beauty_Bathing_Soap (125GM),Personal Care,aa1bbd,STCBE-0,CAMP_SAN_01,50,25.0,25% OFF,37,22
720,P10,Atliq_Cream_Beauty_Bathing_Soap (125GM),Personal Care,31b59d,STVSK-3,CAMP_SAN_01,50,27.0,25% OFF,37,21
721,P10,Atliq_Cream_Beauty_Bathing_Soap (125GM),Personal Care,ad783e,STCHE-4,CAMP_SAN_01,50,30.0,25% OFF,37,25
723,P10,Atliq_Cream_Beauty_Bathing_Soap (125GM),Personal Care,b8c7c8,STHYD-6,CAMP_SAN_01,50,31.0,25% OFF,37,26


4. Identify the product category with the lowest base price before the promotion.
- Personal care is the product category with the lowest base price before the promotion.

In [16]:
df_campaigns=pd.read_csv("dim_campaigns.csv")
df_campaigns.head()

,campaign_id,campaign_name,start_date,end_date
0,CAMP_DIW_01,Diwali,12-11-2023,18-11-2023
1,CAMP_SAN_01,Sankranti,10-01-2024,16-01-2024


In [17]:
df_merged=df_campaigns.merge(df_events1, on='campaign_id')
df_merged.head()

,campaign_id,campaign_name,start_date,end_date,event_id,store_id,product_code,base_price(before_promo),quantity_sold(before_promo),promo_type,base_price(after_promo),quantity_sold(after_promo)
0,CAMP_DIW_01,Diwali,12-11-2023,18-11-2023,8481be,STCHE-1,P04,290,327.0,25% OFF,217,287
1,CAMP_DIW_01,Diwali,12-11-2023,18-11-2023,f30579,STBLR-9,P02,860,337.0,33% OFF,576,488
2,CAMP_DIW_01,Diwali,12-11-2023,18-11-2023,4f570c,STBLR-7,P05,55,122.0,25% OFF,41,107
3,CAMP_DIW_01,Diwali,12-11-2023,18-11-2023,87564a,STCHE-6,P04,290,346.0,25% OFF,217,269
4,CAMP_DIW_01,Diwali,12-11-2023,18-11-2023,6bbadf,STHYD-1,P14,1020,43.0,BOGOF,510,127


In [18]:
df=df_merged[(df_merged['promo_type'] == 'BOGOF') & (df_merged['campaign_name'] == 'Diwali')]
df.head()

,campaign_id,campaign_name,start_date,end_date,event_id,store_id,product_code,base_price(before_promo),quantity_sold(before_promo),promo_type,base_price(after_promo),quantity_sold(after_promo)
4,CAMP_DIW_01,Diwali,12-11-2023,18-11-2023,6bbadf,STHYD-1,P14,1020,43.0,BOGOF,510,127
5,CAMP_DIW_01,Diwali,12-11-2023,18-11-2023,66c422,STBLR-5,P07,300,66.0,BOGOF,150,227
6,CAMP_DIW_01,Diwali,12-11-2023,18-11-2023,108d5a,STCBE-0,P14,1020,29.0,BOGOF,510,101
7,CAMP_DIW_01,Diwali,12-11-2023,18-11-2023,e7d681,STCHE-5,P13,350,66.0,BOGOF,175,226
12,CAMP_DIW_01,Diwali,12-11-2023,18-11-2023,ba86f4,STVSK-1,P13,350,61.0,BOGOF,175,204


In [19]:
df['quantity_sold(after_promo)'].sum()

np.int64(34461)

5. What is the total quantity sold after the promotion for the BOGOF promo type during the Diwali campaign?
- 34461

In [20]:
df1=df_merged[df_merged['campaign_name'] == 'Diwali']
df1.head()

,campaign_id,campaign_name,start_date,end_date,event_id,store_id,product_code,base_price(before_promo),quantity_sold(before_promo),promo_type,base_price(after_promo),quantity_sold(after_promo)
0,CAMP_DIW_01,Diwali,12-11-2023,18-11-2023,8481be,STCHE-1,P04,290,327.0,25% OFF,217,287
1,CAMP_DIW_01,Diwali,12-11-2023,18-11-2023,f30579,STBLR-9,P02,860,337.0,33% OFF,576,488
2,CAMP_DIW_01,Diwali,12-11-2023,18-11-2023,4f570c,STBLR-7,P05,55,122.0,25% OFF,41,107
3,CAMP_DIW_01,Diwali,12-11-2023,18-11-2023,87564a,STCHE-6,P04,290,346.0,25% OFF,217,269
4,CAMP_DIW_01,Diwali,12-11-2023,18-11-2023,6bbadf,STHYD-1,P14,1020,43.0,BOGOF,510,127


In [21]:
df1.groupby('store_id')['quantity_sold(after_promo)'].sum().sort_values(ascending=False).head(1)

store_id
STCHE-4    5013
Name: quantity_sold(after_promo), dtype: int64

6. Which store recorded the highest quantity sold after the promotion during the Diwali campaign?
- STCHE-4    5013

In [22]:
df_merged.head()

,campaign_id,campaign_name,start_date,end_date,event_id,store_id,product_code,base_price(before_promo),quantity_sold(before_promo),promo_type,base_price(after_promo),quantity_sold(after_promo)
0,CAMP_DIW_01,Diwali,12-11-2023,18-11-2023,8481be,STCHE-1,P04,290,327.0,25% OFF,217,287
1,CAMP_DIW_01,Diwali,12-11-2023,18-11-2023,f30579,STBLR-9,P02,860,337.0,33% OFF,576,488
2,CAMP_DIW_01,Diwali,12-11-2023,18-11-2023,4f570c,STBLR-7,P05,55,122.0,25% OFF,41,107
3,CAMP_DIW_01,Diwali,12-11-2023,18-11-2023,87564a,STCHE-6,P04,290,346.0,25% OFF,217,269
4,CAMP_DIW_01,Diwali,12-11-2023,18-11-2023,6bbadf,STHYD-1,P14,1020,43.0,BOGOF,510,127


In [23]:
quantity_before_promo_sankarnti = df_merged.groupby('campaign_name')['quantity_sold(before_promo)'].sum().loc['Sankranti']
quantity_before_promo_sankarnti

np.float64(97894.0)

In [24]:
quantity_after_promo_sankarnti = df_merged.groupby('campaign_name')['quantity_sold(after_promo)'].sum().loc['Sankranti']
quantity_after_promo_sankarnti

np.int64(252069)

In [25]:
quantity_before_promo_diwali = df_merged.groupby('campaign_name')['quantity_sold(before_promo)'].sum().loc['Diwali']
quantity_before_promo_diwali

np.float64(109756.0)

In [26]:
quantity_after_promo_diwali = df_merged.groupby('campaign_name')['quantity_sold(after_promo)'].sum().loc['Diwali']
quantity_after_promo_diwali

np.int64(183404)

In [27]:
sankranti_percent = ((quantity_after_promo_sankarnti-quantity_before_promo_sankarnti)/quantity_before_promo_sankarnti)*100
print(sankranti_percent)

157.49177681982553


In [28]:
diwali_percent = ((quantity_after_promo_diwali-quantity_before_promo_diwali)/quantity_before_promo_diwali)*100
print(diwali_percent)

67.10157075695179


In [29]:
quantity_after_promo_sankarnti-quantity_before_promo_sankarnti

np.float64(154175.0)

7. Understand which campaigns had the most successful outcomes. Compare the total quantities sold before and after the promotions for the Sankranti and Diwali campaigns. Which campaign saw a greater increase in sales?
- Sankranti campaign saw greater increase in sales.

In [30]:
df_merged['revenue_before_promo'] = df_merged['base_price(before_promo)'] * df_merged['quantity_sold(before_promo)']
df_merged['revenue_after_promo'] = df_merged['base_price(after_promo)'] * df_merged['quantity_sold(after_promo)']

In [31]:
df_merged['IR']=(df_merged['revenue_after_promo']-df_merged['revenue_before_promo'])/df_merged['revenue_before_promo']*100

In [32]:
df_merged.head()

,campaign_id,campaign_name,start_date,end_date,event_id,store_id,product_code,base_price(before_promo),quantity_sold(before_promo),promo_type,base_price(after_promo),quantity_sold(after_promo),revenue_before_promo,revenue_after_promo,IR
0,CAMP_DIW_01,Diwali,12-11-2023,18-11-2023,8481be,STCHE-1,P04,290,327.0,25% OFF,217,287,94830.0,62279,-34.325635
1,CAMP_DIW_01,Diwali,12-11-2023,18-11-2023,f30579,STBLR-9,P02,860,337.0,33% OFF,576,488,289820.0,281088,-3.012905
2,CAMP_DIW_01,Diwali,12-11-2023,18-11-2023,4f570c,STBLR-7,P05,55,122.0,25% OFF,41,107,6710.0,4387,-34.619970
3,CAMP_DIW_01,Diwali,12-11-2023,18-11-2023,87564a,STCHE-6,P04,290,346.0,25% OFF,217,269,100340.0,58373,-41.824796
4,CAMP_DIW_01,Diwali,12-11-2023,18-11-2023,6bbadf,STHYD-1,P14,1020,43.0,BOGOF,510,127,43860.0,64770,47.674419


In [33]:
sankranti_df = df_merged[df_merged['campaign_name'] == 'Sankranti'].copy()

product_revenue = sankranti_df.groupby('product_code').agg(
    revenue_before=('revenue_before_promo', 'sum'),
    revenue_after=('revenue_after_promo', 'sum')
).reset_index()

product_revenue['IR%'] = ((product_revenue['revenue_after'] -product_revenue['revenue_before'])/ product_revenue['revenue_before']) * 100

result = product_revenue.sort_values('IR%', ascending=False).head(1)

result

,product_code,revenue_before,revenue_after,IR%
2,P03,3189600.0,6118500,91.826561


In [34]:
result = product_revenue.merge(
    df_products[['product_code', 'product_name']],
    on='product_code',
    how='left'
)

result = result.sort_values('IR%', ascending=False).head(1)

print(result[['product_name', 'IR%']].to_string(index=False))

           product_name       IR%
Atliq_Suflower_Oil (1L) 91.826561


8. Which product recorded the highest Incremental Revenue Percentage (IR%) during the Sankranti campaign? What is the IR% for this product?
   - Atliq_Suflower_Oil (1L),91.83

In [35]:
df_merged['ISU']=(df_merged['quantity_sold(after_promo)']-df_merged['quantity_sold(before_promo)'])/df_merged['quantity_sold(before_promo)']*100
df_merged.head()

,campaign_id,campaign_name,start_date,end_date,event_id,store_id,product_code,base_price(before_promo),quantity_sold(before_promo),promo_type,base_price(after_promo),quantity_sold(after_promo),revenue_before_promo,revenue_after_promo,IR,ISU
0,CAMP_DIW_01,Diwali,12-11-2023,18-11-2023,8481be,STCHE-1,P04,290,327.0,25% OFF,217,287,94830.0,62279,-34.325635,-12.232416
1,CAMP_DIW_01,Diwali,12-11-2023,18-11-2023,f30579,STBLR-9,P02,860,337.0,33% OFF,576,488,289820.0,281088,-3.012905,44.807122
2,CAMP_DIW_01,Diwali,12-11-2023,18-11-2023,4f570c,STBLR-7,P05,55,122.0,25% OFF,41,107,6710.0,4387,-34.619970,-12.295082
3,CAMP_DIW_01,Diwali,12-11-2023,18-11-2023,87564a,STCHE-6,P04,290,346.0,25% OFF,217,269,100340.0,58373,-41.824796,-22.254335
4,CAMP_DIW_01,Diwali,12-11-2023,18-11-2023,6bbadf,STHYD-1,P14,1020,43.0,BOGOF,510,127,43860.0,64770,47.674419,195.348837


In [36]:
df_stores.head()

,store_id,city
0,STTRV-0,Trivandrum
1,STMDU-3,Madurai
2,STHYD-6,Hyderabad
3,STVSK-1,Visakhapatnam
4,STCBE-3,Coimbatore


In [37]:
df_merged = df_merged.merge(
    df_stores[['store_id', 'city']],
    on='store_id',
    how='left'
)

In [38]:
diwali_vizag = df_merged[
    (df_merged['campaign_name'] == 'Diwali') &
    (df_merged['city'] == 'Visakhapatnam')
].copy()

store_isu = diwali_vizag.groupby('store_id').agg(
    units_before=('quantity_sold(before_promo)', 'sum'),
    units_after=('quantity_sold(after_promo)', 'sum')
)

store_isu['ISU%'] = ((store_isu['units_after'] - store_isu['units_before'])/ store_isu['units_before']) * 100

# Lowest ISU%
result = store_isu.sort_values('ISU%').head(1)

result

,units_before,units_after,ISU%
store_id,,,
STVSK-3,1780.0,2656,49.213483


9. Which store in Visakhapatnam recorded the lowest Incremental Sold Units Percentage (ISU%) during the Diwali campaign? What is the ISU% for that store?
    - STVSK-3,49.21


In [39]:
sankranti_df = df_merged[df_merged['campaign_name'] == 'Sankranti'].copy()

promo_summary = sankranti_df.groupby('promo_type').agg(
    revenue_before=('revenue_before_promo', 'sum'),
    revenue_after=('revenue_after_promo', 'sum'),
    units_before=('quantity_sold(before_promo)', 'sum'),
    units_after=('quantity_sold(after_promo)', 'sum')
)

promo_summary['IR%'] = ((promo_summary['revenue_after'] - promo_summary['revenue_before'])/ promo_summary['revenue_before']) * 100

promo_summary['ISU%'] = ((promo_summary['units_after'] - promo_summary['units_before'])/ promo_summary['units_before']) * 100

promo_summary

,revenue_before,revenue_after,units_before,units_after,IR%,ISU%
promo_type,,,,,,
25% OFF,935195.0,567387,6601.0,5307,-39.329552,-19.603090
33% OFF,20483136.0,19273955,33624.0,47459,-5.903300,41.146205
50% OFF,527482.0,363564,4451.0,6100,-31.075563,37.047854
500 Cashback,16185000.0,31027500,5395.0,12411,91.705283,130.046339
BOGOF,19355170.0,36451055,47823.0,180792,88.327227,278.044037


In [40]:
answer = promo_summary[(promo_summary['IR%'] < 0) &(promo_summary['ISU%'] < 0)]
answer

,revenue_before,revenue_after,units_before,units_after,IR%,ISU%
promo_type,,,,,,
25% OFF,935195.0,567387,6601.0,5307,-39.329552,-19.60309


In [41]:
for promo_type, row in answer.iterrows():print(f"{promo_type},{row['IR%']:.2f},{row['ISU%']:.2f}")

25% OFF,-39.33,-19.60


10. Which promo type had both a negative Incremental Revenue Percentage (IR%) and Incremental Sold Units Percentage (ISU%) during the Sankranti campaign?
    - 25% OFF,-39.33,-19.60
